### PR TITLE
[server][da-vinci][admin-tool] Add gRPC streaming ingestion monitor service

### PR DESCRIPTION
**Pull request by sixpluszero**
VALIDATION_OVERRIDE

## Summary
- Add a server-side streaming gRPC service (`VeniceIngestionMonitorService`) that allows continuous monitoring of per-partition ingestion performance with detailed metrics
- The service is **zero-cost when not in use** — per-partition metric collection via `PartitionIngestionMonitor` (LongAdder-based) only activates when a monitoring session starts
- Add `--monitor-ingestion` CLI command to the admin tool for interactive streaming output

> Note: This PR includes a new `.proto` file defining the gRPC service interface. This is not an Avro schema evolution — it is a new gRPC service definition that is integral to the feature implementation. `VALIDATION_OVERRIDE` is used to allow the proto and Java changes in a single PR.

### New gRPC Service
A client connects with a `(version_topic, partition, interval_ms)` request and receives a continuous stream of `IngestionMonitorResponse` messages containing:
- Consumer throughput: records/sec, bytes/sec consumed
- Leader output: records/sec, bytes/sec produced (leader only)
- Latencies: E2E processing, leader preprocessing, leader produce, producer completion, producer callback, storage engine put, value-bytes lookup, RMD lookup
- State: leader/follower state, hybrid flag, elapsed time since last record

### Instrumentation Points (guarded by `if (monitor != null)`)
| File | Metrics |
|------|---------|
| `StoreIngestionTask` | consumed records/bytes, storage put latency, E2E processing latency |
| `LeaderFollowerStoreIngestionTask` | leader preprocessing latency, leader produce latency |
| `LeaderProducerCallback` | leader completion latency, leader callback latency, leader produced bytes |
| `ActiveActiveStoreIngestionTask` | value-bytes lookup latency, RMD lookup latency |

### Concurrency
- One monitoring session per partition (enforced via `ConcurrentHashMap`; duplicate returns `ALREADY_EXISTS`)
- Cleanup on client disconnect via `ServerCallStreamObserver.setOnCancelHandler()`
- Each tick also checks if PCS is still alive (handles partition unsubscribe during monitoring)

### Performance Impact
- **When OFF (default):** single volatile read (`pcs.getIngestionMonitor()`) returning `null` per record (~1-2ns)
- **When ON:** `LongAdder` operations (~10-20ns per call, lock-free), snapshot on separate scheduler thread

### Unit Tests
- `PartitionIngestionMonitorTest` — 7 tests: snapshot/reset semantics, all latency types, concurrent recording accuracy, rate calculations, edge cases
- `VeniceIngestionMonitorServiceImplTest` — 5 tests: successful streaming, error cases (empty topic, non-existent topic/partition), interval clamping

### Integration Test
- `IngestionMonitorGrpcEndToEndTest` — 3 tests: spins up a full Venice cluster with gRPC+SSL, creates a hybrid store, empty push, hooks up the gRPC monitor, streams 100 real-time records, and validates the monitor reports **positive consumed-records rate** (observed ~16 rec/s, 252 bytes/s, 1.3ms E2E latency)

### Admin Tool Usage
```
java -jar venice-admin-tool-all.jar \
  --monitor-ingestion \
  --server-url https://server-host:1690 \
  --store my-store \
  --version 3 \
  --partition 0 \
  --grpc-port 1691 \
  --interval-ms 5000
```

## Test plan
- [x] `PartitionIngestionMonitorTest` — all 7 unit tests pass
- [x] `VeniceIngestionMonitorServiceImplTest` — all 5 unit tests pass
- [x] `ListenerServiceTest` — existing test passes with new parameter
- [x] `IngestionMonitorGrpcEndToEndTest` — all 3 integration tests pass (including positive-rate validation during RT ingestion)
- [x] SpotBugs: `./gradlew :services:venice-server:spotbugsMain` passes clean
- [x] Proto generation: `./gradlew :internal:venice-common:generateProto` succeeds
- [x] Full compilation of `da-vinci-client`, `venice-server`, `venice-admin-tool`, and integration test modules

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+      responseObserver
+          .onError(Status.INVALID_ARGUMENT.withDescription("version_topic is required").asRuntimeException());
+      return;
+    }
+
+    if (intervalMs <= 0) {
+      intervalMs = DEFAULT_INTERVAL_MS;
+    } else if (intervalMs < MIN_INTERVAL_MS) {
+      intervalMs = MIN_INTERVAL_MS;
+    }
```
Comment (sushantmane) [inline services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java:78]: **Missing `MAX_INTERVAL_MS` enforcement.** The server clamps the minimum but has no upper bound. A client could request `intervalMs = Integer.MAX_VALUE`, holding a session slot and a scheduler thread indefinitely while producing no useful output.

Add an upper bound (e.g., 10 minutes):
```java
private static final int MAX_INTERVAL_MS = 600_000; // 10 minutes

if (intervalMs <= 0) {
    intervalMs = DEFAULT_INTERVAL_MS;
} else {
    intervalMs = Math.max(MIN_INTERVAL_MS, Math.min(intervalMs, MAX_INTERVAL_MS));
}
```

Also update the proto comment to document the max:
```proto
int32 interval_ms = 3; // default 5000, min 1000, max 600000
```

File: services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+public class VeniceIngestionMonitorServiceImpl
+    extends VeniceIngestionMonitorServiceGrpc.VeniceIngestionMonitorServiceImplBase implements Closeable {
+  private static final Logger LOGGER = LogManager.getLogger(VeniceIngestionMonitorServiceImpl.class);
+  private static final int MIN_INTERVAL_MS = 1000;
+  private static final int DEFAULT_INTERVAL_MS = 5000;
+
+  private final KafkaStoreIngestionService ingestionService;
+  private final ConcurrentHashMap<String, ActiveSession> activeSessions = new ConcurrentHashMap<>();
+  private final ScheduledExecutorService scheduler =
+      Executors.newScheduledThreadPool(2, r -> new Thread(r, "ingestion-monitor-scheduler"));
```
Comment (sushantmane) [inline services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java]: **Use `DaemonThreadFactory` instead of raw `Thread` construction.** The rest of the Venice codebase consistently uses `DaemonThreadFactory` for thread pool creation (e.g., `RouterServer`, `DeferredVersionSwapService`, etc.). This ensures daemon threads (so the pool doesn't prevent JVM shutdown) and integrates with `LogContext` for proper log correlation.

```java
private final ScheduledExecutorService scheduler =
    Executors.newScheduledThreadPool(2, new DaemonThreadFactory("ingestion-monitor-scheduler"));
```

Also consider making the pool size configurable rather than hardcoded to 2. A production server hosting many partitions with active monitoring sessions could see timer drift with only 2 threads.

File: services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+      responseObserver.onError(
+          Status.NOT_FOUND
+              .withDescription("No partition consumption state for partition " + partition + " of " + versionTopic)
+              .asRuntimeException());
+      return;
+    }
+
+    // Create the monitor and atomically register the session to prevent duplicates
+    PartitionIngestionMonitor monitor = new PartitionIngestionMonitor();
+    ActiveSession newSession = new ActiveSession(null, monitor, pcs);
```
Comment (sushantmane) [inline services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java:114]: **Race condition: `future` is `null` when the first scheduled tick fires.**

The session is registered with `future = null`, then `scheduleAtFixedRate` starts immediately (initial delay = 0). The first tick can fire and complete *before* `setFuture()` on line 169 is called.

If that first tick finds PCS is gone, it calls `cleanup()` → `session.getFuture()` returns **null** → the future is never cancelled. Then `setFuture(future)` sets the future on an already-removed session — **orphaned recurring task / resource leak**.

**Suggested fix:** Either:
1. Use a non-zero initial delay equal to `intervalMs`, and emit the first snapshot synchronously (inline) before scheduling.
2. Or use an `AtomicReference<ScheduledFuture<?>>` that the cleanup lambda can cancel-after-set.

File: services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+              .debug("Failed to send error to observer for {} (likely already closed)", sessionKey, observerException);
+        }
+      }
+    }, 0, intervalMs, TimeUnit.MILLISECONDS);
+
+    // Update the session with the scheduled future
+    newSession.setFuture(future);
+
+    // Handle client disconnect
+    serverObserver.setOnCancelHandler(() -> {
```
Comment (sushantmane) [inline services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java]: **`setOnCancelHandler` is registered *after* scheduling starts.** The scheduled task fires immediately (initial delay = 0) and the client could disconnect between `scheduleAtFixedRate` and this line. If that happens, the cancel handler never fires.

The `isCancelled()` check inside the scheduled task provides a fallback, but there's an unnecessary gap. Move `setOnCancelHandler` **before** `scheduleAtFixedRate` to close this window.

File: services/venice-server/src/main/java/com/linkedin/venice/listener/ListenerService.java [hunk 5]
```diff
@@ -173,6 +176,9 @@ public ListenerService(
           .setExecutor(grpcExecutor)
           .setInterceptors(interceptors);
 
+      kafkaStoreIngestionService
+          .ifPresent(service -> grpcServerBuilder.addService(new VeniceIngestionMonitorServiceImpl(service)));
```
Comment (sushantmane) [inline services/venice-server/src/main/java/com/linkedin/venice/listener/ListenerService.java]: **`VeniceIngestionMonitorServiceImpl.close()` is never called — resource leak.**

The service instance is created inline and passed into `grpcServerBuilder.addService()`, so no reference is retained. When the gRPC server shuts down, it cancels in-flight RPCs but does **not** call `close()` on registered services. The internal `ScheduledExecutorService` and any active sessions will leak.

Store a reference and call `close()` in `ListenerService.stopInner()`, or register a gRPC server lifecycle listener.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/ActiveActiveStoreIngestionTask.java [hunk 6]
```diff
@@ -413,9 +413,17 @@ byte[] getRmdWithValueSchemaByteBufferFromStorage(
     final long lookupStartTimeInNS = System.nanoTime();
     ValueRecord result = databaseLookupWithConcurrencyLimit(
         () -> getRmdWithValueSchemaByteBufferFromStorageInternal(partition, key, rmdManifestContainer));
+    long rmdLookupElapsedNs = System.nanoTime() - lookupStartTimeInNS;
```
Comment (sushantmane) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/ActiveActiveStoreIngestionTask.java:487]: **Duplicate `System.nanoTime()` call on the hot path.** This line computes the elapsed time in ns, then the existing `recordIngestionReplicationMetadataLookUpLatency` call below it calls `LatencyUtils.getElapsedTimeFromNSToMS(lookupStartTimeInNS)` which does another `System.nanoTime()` internally (~25ns each call).

Compute once and share:
```java
long endNs = System.nanoTime();
long elapsedNs = endNs - lookupStartTimeInNS;
getHostLevelIngestionStats().recordIngestionReplicationMetadataLookUpLatency(
    elapsedNs / 1_000_000.0, currentTimeForMetricsMs);
// ... monitor.recordRmdLookupLatencyNs(elapsedNs);
```

Same pattern applies to `getValueBytesForKey()` (line ~802) and the storage put / E2E latency sites in `StoreIngestionTask`.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderProducerCallback.java [hunk 7]
```diff
@@ -222,6 +226,14 @@ public void onCompletion(PubSubProduceResult produceResult, Exception e) {
                   LatencyUtils.getElapsedTimeFromMsToMs(currentTimeForMetricsMs),
                   currentTimeForMetricsMs);
         }
+        PartitionIngestionMonitor callbackMonitor = partitionConsumptionState.getIngestionMonitor();
+        if (callbackMonitor != null) {
+          // Measures time spent in callback processing (drainer queuing, chunk handling, etc.)
+          // Same measurement as recordProducerCallBackLatency above, converted to ns for precision
+          callbackMonitor
+              .recordLeaderCallbackLatencyNs((System.currentTimeMillis() - currentTimeForMetricsMs) * 1_000_000L);
```
Comment (sushantmane) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderProducerCallback.java]: **Wall-clock millisecond precision for a latency metric.** All other latency metrics in this PR use `System.nanoTime()` for nanosecond precision, but this callback latency uses:
```java
(System.currentTimeMillis() - currentTimeForMetricsMs) * 1_000_000L
```
`currentTimeMillis()` resolution is typically 10-16ms on most JVMs, so sub-millisecond callbacks will show as 0ns. Consider capturing `System.nanoTime()` at callback entry for consistency, or at least document why this one differs.

File: services/venice-server/src/test/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        .setVersionTopic(VERSION_TOPIC)
+        .setPartition(PARTITION)
+        .setIntervalMs(100) // below minimum of 1000
+        .build();
+
+    Iterator<IngestionMonitorResponse> responses = blockingStub.monitorIngestion(request);
+    assertTrue(responses.hasNext());
+    // Should still work, just with clamped interval
+    assertNotNull(responses.next());
+  }
```
Comment (sushantmane) [inline services/venice-server/src/test/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImplTest.java:182]: **Missing test for `ALREADY_EXISTS` duplicate-session behavior.** The service guarantees one session per partition via `putIfAbsent`, which is an important concurrency invariant. Consider adding a test that opens a monitoring session, then attempts a second one for the same partition, and asserts `ALREADY_EXISTS` status.

**Review by sushantmane — COMMENTED** [review]
Thorough review focusing on concurrency, efficiency, duplication, and the missing max-interval validation. Overall a well-designed feature — the zero-cost-when-off semantics and LongAdder-based approach are solid choices.

File: services/venice-server/src/main/java/com/linkedin/venice/listener/ListenerService.java [hunk 9]
```diff
@@ -211,6 +220,9 @@ public void stopInner() throws Exception {
     bossGroup.shutdownGracefully();
     shutdown.sync();
 
+    if (ingestionMonitorService != null) {
+      ingestionMonitorService.close();
+    }
     if (grpcServer != null) {
       LOGGER.info("Stopping gRPC service on port {}", grpcPort);
       grpcServer.stop();
```
[the remark this answers is not in the document]
Comment (sixpluszero) [reply services/venice-server/src/main/java/com/linkedin/venice/listener/ListenerService.java:228]: Good catch — fixed in 8d88c6a. Added a `closed` AtomicBoolean to the service: `close()` flips it before iterating active sessions, the `monitorIngestion` entry point rejects new calls early with `UNAVAILABLE`, and `scheduleAtFixedRate` is wrapped in a try/catch so any call that slips past the early check self-cleans on `RejectedExecutionException` instead of leaking a session/monitor.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 10]
```diff
@@ -5231,13 +5239,18 @@ private int processKafkaDataMessage(
...
-          double putLatency = LatencyUtils.getElapsedTimeFromNSToMS(startTimeNs);
+          double putLatency = putElapsedNs / 1_000_000.0;
           versionedIngestionStats.recordStorageEnginePutTime(storeName, versionNumber, putLatency);
           if (tehutiRecordMetrics) {
             hostLevelIngestionStats.recordStorageEnginePutLatency(putLatency, currentTimeMs);
           }
         }
+        PartitionIngestionMonitor putMonitor = partitionConsumptionState.getIngestionMonitor();
+        if (putMonitor != null) {
+          putMonitor.recordStoragePutLatencyNs(putElapsedNs);
```
[the remark this answers is not in the document]
Comment (sixpluszero) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java]: Good catch — fixed in 8d88c6a. Hoisted the `PartitionIngestionMonitor` lookup in `processKafkaDataMessage` and added `ingestionMonitor != null` to the `startTimeNs` gate, so the monitor records a real `putElapsedNs` even when record-level metrics and trace logging are off. (The hoisted variable also replaces the duplicate `getIngestionMonitor()` call at the original line 5250.)

File: services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+      if (session != null) {
+        ScheduledFuture<?> future = session.getFuture();
+        if (future != null) {
+          future.cancel(false);
+        }
+        if (session.pcs.getIngestionMonitor() == session.monitor) {
+          session.pcs.setIngestionMonitor(null);
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (sixpluszero) [reply services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java]: Good catch — fixed in 8d88c6a. `cleanup` now takes the expected `ActiveSession` and uses `activeSessions.remove(key, expectedSession)` (and reads `pcs`/`monitor` off the session itself), so a stale tick from an already-cancelled session can no longer evict — or detach the monitor of — a newer session that reused the same topic-partition key.

**Comment by sushantmane** [comment]
### Code review

Found 1 issue:

1. **Concurrent `onNext()` calls from the 2-thread scheduler pool can corrupt the gRPC stream** (bug in `VeniceIngestionMonitorServiceImpl`)

`scheduleAtFixedRate` with a 2-thread pool allows two ticks of the **same session** to overlap if a tick runs longer than `intervalMs`. When the client reads slowly (gRPC flow control backpressure), `onNext()` blocks on the write. If the block exceeds `intervalMs` (minimum 1000 ms), the next scheduled invocation fires on the second thread and calls `onNext()` concurrently on the same `ServerCallStreamObserver`. The gRPC contract explicitly states that `StreamObserver` instances are **not thread-safe** — concurrent calls produce undefined behavior (stream corruption or IllegalStateException).

The fix is to replace `scheduleAtFixedRate` with `scheduleWithFixedDelay` so the next tick only starts after the current one completes, or to use a single-thread pool (`newSingleThreadScheduledExecutor`).

https://github.com/linkedin/venice/blob/8d88c6ae0820d5c56b50a6657e22154bfe8be528/services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java#L44-L48

https://github.com/linkedin/venice/blob/8d88c6ae0820d5c56b50a6657e22154bfe8be528/services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java#L126-L175

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+    pcs.setIngestionMonitor(monitor);
+
+    ServerCallStreamObserver<IngestionMonitorResponse> serverObserver =
+        (ServerCallStreamObserver<IngestionMonitorResponse>) responseObserver;
+
+    // Register cancel handler before scheduling to avoid missing early disconnects
+    serverObserver.setOnCancelHandler(() -> {
+      LOGGER.info("Client disconnected from monitoring session for {}", sessionKey);
+      cleanup(sessionKey, pcs);
+    });
```
[the remark this answers is not in the document]
Comment (sixpluszero) [reply services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java:139]: Fixed in 3ee7afb. The `ScheduledFuture` is published to the session via `setFuture()`, then we immediately re-check `activeSessions.get(sessionKey) != newSession`. If the cancel handler already ran `cleanup()` before the future existed (so `cleanup()` saw a null future and couldn't cancel it), the session is no longer registered and we cancel the now-published future right there — the periodic task can no longer keep ticking against a detached session.

File: services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+    if (existing != null) {
+      responseObserver.onError(
+          Status.ALREADY_EXISTS.withDescription("A monitoring session is already active for " + sessionKey)
+              .asRuntimeException());
+      return;
+    }
+
+    // Attach monitor to PCS now that we own the session
+    pcs.setIngestionMonitor(monitor);
+
```
[the remark this answers is not in the document]
Comment (sixpluszero) [reply services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java:134]: Fixed in 3ee7afb. The `ServerCallStreamObserver` cast is now guarded by an `instanceof` check performed *before* the session is registered or the monitor attached; if the cast would fail, the call ends with `INTERNAL` and nothing is leaked. Covered by the new `testNonServerCallStreamObserverFailsWithInternal`.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderProducerCallback.java [hunk 14]
```diff
@@ -90,6 +90,7 @@ public void onCompletion(PubSubProduceResult produceResult, Exception e) {
             e);
       }
     } else {
+      long callbackStartNs = System.nanoTime();
```
[the remark this answers is not in the document]
Comment (sixpluszero) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderProducerCallback.java]: Fixed in 3ee7afb. The monitor is looked up once at the top of the success branch and `callbackStartNs` is only captured when it is non-null, so the success path makes no extra `System.nanoTime()` call when monitoring is off. The single reference is reused for both the completion- and callback-latency metrics.

File: clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/PartitionIngestionMonitorTest.java [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+
+    assertTrue(latch.await(10, TimeUnit.SECONDS), "Timed out waiting for concurrent recording");
+    executor.shutdown();
+
+    PartitionIngestionSnapshot snapshot = monitor.snapshotAndReset(1000);
+
+    long expectedRecords = (long) numThreads * recordsPerThread;
+    long expectedBytes = expectedRecords * 10;
+
+    // Due to LongAdder's eventual consistency, values should be exact
```
[the remark this answers is not in the document]
Comment (sixpluszero) [reply clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/PartitionIngestionMonitorTest.java]: Fixed in 3ee7afb. Reworded to: `// Values should be exact: the snapshot is taken after the latch, with no concurrent updates in flight.`

File: services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+        ScheduledFuture<?> future = session.getFuture();
+        if (future != null) {
+          future.cancel(false);
+        }
+        if (session.pcs.getIngestionMonitor() == session.monitor) {
+          session.pcs.setIngestionMonitor(null);
+        }
+      }
+    }
+    scheduler.shutdownNow();
```
[the remark this answers is not in the document]
Comment (sixpluszero) [reply services/venice-server/src/main/java/com/linkedin/venice/listener/grpc/VeniceIngestionMonitorServiceImpl.java:265]: Fixed in 3ee7afb. Two changes close this window: (1) `monitorIngestion` re-checks `closed` immediately after `putIfAbsent` and, if shutdown has started, removes its own session and returns `UNAVAILABLE` — so a session can't be attached past the flag; (2) `close()` now drains `activeSessions` in a loop until empty instead of a single `keySet()` pass, catching any session registered concurrently with the drain. Covered by the new `testMonitorIngestionRejectedAfterClose`.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 17]
```diff
@@ -5231,13 +5241,17 @@ private int processKafkaDataMessage(
...
         if (recordMetrics) {
-          double putLatency = LatencyUtils.getElapsedTimeFromNSToMS(startTimeNs);
+          double putLatency = putElapsedNs / 1_000_000.0;
           versionedIngestionStats.recordStorageEnginePutTime(storeName, versionNumber, putLatency);
           if (tehutiRecordMetrics) {
             hostLevelIngestionStats.recordStorageEnginePutLatency(putLatency, currentTimeMs);
           }
         }
+        if (ingestionMonitor != null) {
+          ingestionMonitor.recordStoragePutLatencyNs(putElapsedNs);
```
[the remark this answers is not in the document]
Comment (sixpluszero) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java]: Fixed in 3ee7afb. The `putElapsedNs = System.nanoTime() - startTimeNs` computation is now inside an `if (recordMetrics || ingestionMonitor != null)` guard, so the PUT path makes no `nanoTime()` call when neither record metrics nor a monitor are active; behavior for enabled metrics is unchanged.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderProducerCallback.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (sixpluszero) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderProducerCallback.java]: Fixed in 3ee7afb (same change as the related callback thread). The `PartitionIngestionMonitor` is looked up once at the start of the success branch and `callbackStartNs` is only captured when it is non-null, with that single reference reused for the callback metrics — so there is no `System.nanoTime()` on the hot path when no session is active.

File: services/venice-server/src/main/java/com/linkedin/venice/listener/ListenerService.java [hunk 18]
```diff
@@ -173,6 +177,11 @@ public ListenerService(
           .setExecutor(grpcExecutor)
           .setInterceptors(interceptors);
 
+      kafkaStoreIngestionService.ifPresent(service -> {
+        ingestionMonitorService = new VeniceIngestionMonitorServiceImpl(service);
+        grpcServerBuilder.addService(ingestionMonitorService);
```
[the remark this answers is not in the document]
Comment (sixpluszero) [reply services/venice-server/src/main/java/com/linkedin/venice/listener/ListenerService.java:182]: Fixed in 3ee7afb. Both gRPC ACL interceptors (`ServerAclHandler` and `ServerStoreAclHandler`) now branch on message type: the `VeniceClientRequest`-based read ACL runs only for read-service messages, and any other message (e.g. `IngestionMonitorRequest`) is forwarded instead of cast — eliminating the `ClassCastException`. As a result the monitor endpoint is gated by mutual TLS (`VerifySslHandler`) rather than store-level read ACL, consistent with the existing non-data diagnostic endpoints; it exposes ingestion-rate/latency telemetry, not store data/keys/values. If we'd prefer per-store authorization here, the follow-up is to derive the store from `version_topic` and run a store read-ACL check — happy to do that if you want to enforce that posture. Pass-through is covered by new tests in `ServerAclHandlerTest`/`ServerStoreAclHandlerTest`.

File: clients/venice-admin-tool/src/main/java/com/linkedin/venice/AdminTool.java [hunk 19]
```diff
@@ -3509,6 +3512,121 @@ private static void dumpTopicPartitionIngestionContext(CommandLine cmd) throws E
...
+    }
+
+    // Extract host from server URL
+    String host;
+    try {
+      java.net.URI uri = new java.net.URI(serverUrl);
+      host = uri.getHost();
+    } catch (Exception e) {
+      throw new VeniceException("Invalid server URL: " + serverUrl, e);
+    }
```
[the remark this answers is not in the document]
Comment (sixpluszero) [reply clients/venice-admin-tool/src/main/java/com/linkedin/venice/AdminTool.java:3553]: Fixed in 023a380. `host` is now validated after parsing — if `new URI(serverUrl).getHost()` returns null/empty (e.g. a scheme-less input like `localhost:1690`, where `localhost` is parsed as the URI scheme), we throw a clear `VeniceException` asking for a URL with an explicit scheme and host instead of building a `null:<port>` target that fails with a confusing gRPC error. Also narrowed the catch to `URISyntaxException` as suggested.

**Review by sushantmane — APPROVED** [review]
Thanks!
